service/lib/dinstaller/software/proposal.rb

```diff
@@ -0,0 +1,211 @@
+      #
+      # @return [Integer] size of the installation in bytes
+      def packages_size
+        Yast::Pkg.PkgMediaSizes.reduce(0) do |res, media_size|
+          media_size.reduce(res, :+)
+        end
+      end
+
+      # Determines whether the proposal is valid
+      #
+      # @return [Boolean]
+      def valid?
+        @calculated && @errors.empty?
+      end
+
+    private
+
+      # @return [Logger]
+      attr_reader :logger
+
+      # Initializes the target, closing the previous one
+      def initialize_target
+        Yast::Pkg.TargetFinish # ensure that previous target is closed
+        Yast::Pkg.TargetInitialize(Yast::Installation.destdir)
+        Yast::Pkg.TargetLoad
+        Yast::Pkg.SetAdditionalLocales(languages)
+        Yast::Pkg.SetSolverFlags("ignoreAlreadyRecommended" => false, "onlyRequires" => true)
+      end
+
+      # Selects the base product
+      #
+      # @see #base_product
+      def select_base_product
+        base_product = Y2Packager::Product.available_base_products.find do |product|
+          product.name == @base_product
+        end
+
+        if base_product.nil?
+          logger.error "Could not select the base product '#{@base_product}'"
+        else
+          logger.info "Selecting the base product '#{base_product.name}'"
+          base_product&.select
+        end
+      end
+
+      # Returns the errors from the attempt to create a proposal
+      #
+      # It collects errors from:
+      #
+      # * The proposal result
+      # * The last solver execution
+      #
+      # @param proposal_result [Hash] Proposal result; it might contain a "warning" key with warning
+      #   messages.
+      # @return [Array<ValidationError>] List of errors
+      def find_errors(proposal_result)
+        msgs = []
+        msgs.concat(warning_messages(proposal_result))
+        msgs.concat(solver_messages)
+        msgs.map { |m| ValidationError.new(m) }
+      end
+
+      # Runs the solver to satisfy the solve_dependencies
+      #
+      # If the solver failed, it logs the error.
+      #
+      # @return [Boolean] whether the solver ran successfully
+      def solve_dependencies
+        res = Yast::Pkg.PkgSolve(unused = true)
+        logger.info "Solver run #{res.inspect}"
+        return true if res
+
+        logger.error "Solver failed: #{Yast::Pkg.LastError}"
+        logger.error "Details: #{Yast::Pkg.LastErrorDetails}"
+        logger.error "Solver errors: #{Yast::Pkg.PkgSolveErrors}"
+        false
+      end
+
+      # Extracts the warning messages from the proposal result
+      #
+      # @param proposal_result [Hash] Proposal result; it might contain a "warning" key with warning
+      #   messages.
+      def warning_messages(proposal_result)
+        return [] unless proposal_result["warning_level"] == :blocker
+
+        proposal_result["warning"]
+          .split("<br>")
+          .grep_v(/Please manually select .*/) # FIXME: it depends on the language
```

Comment:
Should we create a card for this FIXME?

Reply:
Yes. Actually, we need to go through all the FIXMEs and TODOs...